File: src/market.js
// ============================================================
// MARKET DATA — Yahoo Finance (free, no API key required)
// Routes through /api/yahoo Vite dev proxy to bypass browser CORS,
// with a public CORS proxy fallback for non-dev environments.
// ============================================================

const YAHOO_HOST  = 'https://query1.finance.yahoo.com';
const CORS_PROXY  = 'https://corsproxy.io/?';

// Ticker → Yahoo symbol overrides for indices, currencies, commodities.
const SPECIAL_MAP = {
  NIFTY:      '^NSEI',
  NIFTY50:    '^NSEI',
  SENSEX:     '^BSESN',
  BANKNIFTY:  '^NSEBANK',
  INDIAVIX:   '^INDIAVIX',
  NIFTYIT:    '^CNXIT',
  NIFTYBANK:  '^NSEBANK',
  NIFTYAUTO:  '^CNXAUTO',
  NIFTYPHARMA:'^CNXPHARMA',
  DOW:        '^DJI',
  DOWJONES:   '^DJI',
  SP500:      '^GSPC',
  NASDAQ:     '^IXIC',
  USDINR:     'INR=X',
  CRUDE:      'CL=F',
  GOLD:       'GC=F',
};

export function toYahooSymbol(ticker) {
  if (!ticker) return '';
  const raw = String(ticker).toUpperCase().trim();
  if (raw.startsWith('^')) return raw;
  const key = raw.replace(/\s+/g, '').replace(/[-_&]/g, '');
  if (SPECIAL_MAP[key]) return SPECIAL_MAP[key];
  if (raw.includes('=') || raw.endsWith('.NS') || raw.endsWith('.BO')) return raw;
  // Yahoo uses "M_M.NS" for M&M; strip & from source to match.
  const cleaned = raw.replace(/&/g, '_');
  return `${cleaned}.NS`;
}

// Curated NSE constituent lists — used by Today's Picks so we can render
// real prices without an LLM call. Update annually if index composition shifts.
export const SECTOR_TICKERS = {
  largecap: [
    { ticker: 'RELIANCE',    name: 'Reliance Industries' },
    { ticker: 'HDFCBANK',    name: 'HDFC Bank' },
    { ticker: 'TCS',         name: 'Tata Consultancy Services' },
    { ticker: 'INFY',        name: 'Infosys' },
    { ticker: 'ICICIBANK',   name: 'ICICI Bank' },
    { ticker: 'BHARTIARTL',  name: 'Bharti Airtel' },
    { ticker: 'ITC',         name: 'ITC' },
    { ticker: 'LT',          name: 'Larsen & Toubro' },
    { ticker: 'SBIN',        name: 'State Bank of India' },
    { ticker: 'HINDUNILVR',  name: 'Hindustan Unilever' },
    { ticker: 'BAJFINANCE',  name: 'Bajaj Finance' },
    { ticker: 'MARUTI',      name: 'Maruti Suzuki' },
    { ticker: 'ASIANPAINT',  name: 'Asian Paints' },
    { ticker: 'AXISBANK',    name: 'Axis Bank' },
    { ticker: 'KOTAKBANK',   name: 'Kotak Mahindra Bank' },
    { ticker: 'SUNPHARMA',   name: 'Sun Pharma' },
    { ticker: 'TITAN',       name: 'Titan Company' },
    { ticker: 'ULTRACEMCO',  name: 'UltraTech Cement' },
    { ticker: 'ADANIENT',    name: 'Adani Enterprises' },
    { ticker: 'NTPC',        name: 'NTPC' },
  ],
  midcap: [
    { ticker: 'PERSISTENT',  name: 'Persistent Systems' },
    { ticker: 'COFORGE',     name: 'Coforge' },
    { ticker: 'DIXON',       name: 'Dixon Technologies' },
    { ticker: 'POLYCAB',     name: 'Polycab India' },
    { ticker: 'MPHASIS',     name: 'Mphasis' },
    { ticker: 'INDIGO',      name: 'InterGlobe Aviation' },
    { ticker: 'ASTRAL',      name: 'Astral' },
    { ticker: 'BALKRISIND',  name: 'Balkrishna Industries' },
    { ticker: 'AUBANK',      name: 'AU Small Finance Bank' },
    { ticker: 'CUMMINSIND',  name: 'Cummins India' },
    { ticker: 'PIIND',       name: 'PI Industries' },
    { ticker: 'JUBLFOOD',    name: 'Jubilant FoodWorks' },
    { ticker: 'TVSMOTOR',    name: 'TVS Motor' },
    { ticker: 'MRF',         name: 'MRF' },
  ],
  it: [
    { ticker: 'TCS',         name: 'Tata Consultancy Services' },
    { ticker: 'INFY',        name: 'Infosys' },
    { ticker: 'HCLTECH',     name: 'HCL Technologies' },
    { ticker: 'WIPRO',       name: 'Wipro' },
    { ticker: 'TECHM',       name: 'Tech Mahindra' },
    { ticker: 'LTIM',        name: 'LTIMindtree' },
    { ticker: 'PERSISTENT',  name: 'Persistent Systems' },
    { ticker: 'COFORGE',     name: 'Coforge' },
    { ticker: 'MPHASIS',     name: 'Mphasis' },
    { ticker: 'OFSS',        name: 'Oracle Financial Services' },
  ],
  bank: [
    { ticker: 'HDFCBANK',    name: 'HDFC Bank' },
    { ticker: 'ICICIBANK',   name: 'ICICI Bank' },
    { ticker: 'SBIN',        name: 'State Bank of India' },
    { ticker: 'KOTAKBANK',   name: 'Kotak Mahindra Bank' },
    { ticker: 'AXISBANK',    name: 'Axis Bank' },
    { ticker: 'INDUSINDBK',  name: 'IndusInd Bank' },
    { ticker: 'FEDERALBNK',  name: 'Federal Bank' },
    { ticker: 'IDFCFIRSTB',  name: 'IDFC First Bank' },
    { ticker: 'PNB',         name: 'Punjab National Bank' },
    { ticker: 'BANKBARODA',  name: 'Bank of Baroda' },
  ],
  auto: [
    { ticker: 'MARUTI',      name: 'Maruti Suzuki' },
    { ticker: 'M&M',         name: 'Mahindra & Mahindra' },
    { ticker: 'TATAMOTORS',  name: 'Tata Motors' },
    { ticker: 'BAJAJ-AUTO',  name: 'Bajaj Auto' },
    { ticker: 'HEROMOTOCO',  name: 'Hero MotoCorp' },
    { ticker: 'EICHERMOT',   name: 'Eicher Motors' },
    { ticker: 'TVSMOTOR',    name: 'TVS Motor' },
    { ticker: 'ASHOKLEY',    name: 'Ashok Leyland' },
    { ticker: 'MRF',         name: 'MRF' },
    { ticker: 'BHARATFORG',  name: 'Bharat Forge' },
  ],
  pharma: [
    { ticker: 'SUNPHARMA',   name: 'Sun Pharma' },
    { ticker: 'DRREDDY',     name: 'Dr. Reddys Labs' },
    { ticker: 'CIPLA',       name: 'Cipla' },
    { ticker: 'DIVISLAB',    name: 'Divis Laboratories' },
    { ticker: 'LUPIN',       name: 'Lupin' },
    { ticker: 'AUROPHARMA',  name: 'Aurobindo Pharma' },
    { ticker: 'TORNTPHARM',  name: 'Torrent Pharma' },
    { ticker: 'ZYDUSLIFE',   name: 'Zydus Lifesciences' },
    { ticker: 'BIOCON',      name: 'Biocon' },
    { ticker: 'GLENMARK',    name: 'Glenmark Pharma' },
  ],
};

// momentum & value are computed dynamically from largecap by sort/filter
SECTOR_TICKERS.momentum = SECTOR_TICKERS.largecap;
SECTOR_TICKERS.value    = SECTOR_TICKERS.largecap;

async function yahooFetch(path) {
  // 1) Try Vite dev proxy (fastest, most reliable in dev)
  try {
    const res = await fetch(`/api/yahoo${path}`);
    if (res.ok) {
      const ct = res.headers.get('content-type') || '';
      if (ct.includes('application/json')) return res.json();
    }
  } catch { /* fall through */ }

  // 2) Fallback: public CORS proxy (works in preview/prod static)
  const url = `${CORS_PROXY}${encodeURIComponent(YAHOO_HOST + path)}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`Yahoo ${res.status}`);
  return res.json();
}

export async function getQuote(ticker) {
  const symbol = toYahooSymbol(ticker);
  const data = await yahooFetch(`/v8/finance/chart/${encodeURIComponent(symbol)}?range=5d&interval=1d`);
  const result = data?.chart?.result?.[0];
  if (!result) throw new Error(`No data for ${ticker}`);
  const meta  = result.meta || {};
  const price = meta.regularMarketPrice;
  const prev  = meta.chartPreviousClose ?? meta.previousClose ?? price;
  const change    = price - prev;
  const changePct = prev ? (change / prev) * 100 : 0;
  return {
    ticker: String(ticker).toUpperCase(),
    symbol,
    name:  meta.longName || meta.shortName || String(ticker).toUpperCase(),
    price,
    previousClose: prev,
    change,
    changePct,
    currency: meta.currency || 'INR',
    dayHigh:  meta.regularMarketDayHigh,
    dayLow:   meta.regularMarketDayLow,
    volume:   meta.regularMarketVolume,
    exchange: meta.exchangeName,
  };
}

export async function getHistory(ticker, range = '5y', interval = '1mo') {
  const symbol = toYahooSymbol(ticker);
  const data = await yahooFetch(`/v8/finance/chart/${encodeURIComponent(symbol)}?range=${range}&interval=${interval}`);
  const result = data?.chart?.result?.[0];
  if (!result) throw new Error(`No history for ${ticker}`);
  const timestamps = result.timestamp || [];
  const closes = result.indicators?.quote?.[0]?.close || [];
  return timestamps
    .map((t, i) => ({ date: new Date(t * 1000), close: closes[i] }))
    .filter(p => p.close != null);
}

export function computeReturns(history) {
  if (!history || history.length < 2) return null;
  const last = history[history.length - 1];
  const price = last.close;
  const today = last.date;
  const daysBefore = (d) => new Date(today.getTime() - d * 86400000);
  const findAtOrBefore = (target) => {
    let best = null;
    for (const p of history) {
      if (p.date <= target) best = p; else break;
    }
    return best;
  };
  const pct  = (past) => past?.close ? ((price - past.close) / past.close) * 100 : null;
  const cagr = (past, years) => past?.close ? (Math.pow(price / past.close, 1 / years) - 1) * 100 : null;
  const fmt  = (n, suffix = '') => n == null ? null : `${n >= 0 ? '+' : ''}${n.toFixed(1)}%${suffix}`;

  return {
    '1M': fmt(pct(findAtOrBefore(daysBefore(30)))),
    '3M': fmt(pct(findAtOrBefore(daysBefore(90)))),
    '6M': fmt(pct(findAtOrBefore(daysBefore(180)))),
    '1Y': fmt(pct(findAtOrBefore(daysBefore(365)))),
    '3Y': fmt(cagr(findAtOrBefore(daysBefore(365 * 3)), 3), ' CAGR'),
    '5Y': fmt(cagr(findAtOrBefore(daysBefore(365 * 5)), 5), ' CAGR'),
  };
}

export async function getBatchQuotes(tickers) {
  const results = await Promise.allSettled(tickers.map(t => getQuote(t)));
  return results.map((r, i) =>
    r.status === 'fulfilled'
      ? r.value
      : { ticker: tickers[i], error: r.reason?.message, failed: true }
  );
}

export function formatPrice(price, currency = 'INR') {
  if (price == null || !Number.isFinite(price)) return '—';
  const sym = currency === 'INR' ? '₹' : currency === 'USD' ? '$' : `${currency} `;
  return `${sym}${price.toLocaleString('en-IN', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
}

export function formatChangePct(pct) {
  if (pct == null || !Number.isFinite(pct)) return '—';
  return `${pct >= 0 ? '+' : ''}${pct.toFixed(2)}%`;
}

File: backend/app/market.py
"""Yahoo Finance client. Mirrors src/market.js so the frontend can migrate
to backend calls without behaviour changes."""
import logging
import re
from typing import Any, Dict, List, Optional

import httpx
from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type

from .config import get_settings

log = logging.getLogger("desk.market")

# NSE / index / commodity aliases — same map as market.js.
SPECIAL_MAP: Dict[str, str] = {
    "NIFTY": "^NSEI", "NIFTY50": "^NSEI",
    "SENSEX": "^BSESN", "BANKNIFTY": "^NSEBANK", "INDIAVIX": "^INDIAVIX",
    "NIFTYIT": "^CNXIT", "NIFTYBANK": "^NSEBANK",
    "NIFTYAUTO": "^CNXAUTO", "NIFTYPHARMA": "^CNXPHARMA",
    "DOW": "^DJI", "DOWJONES": "^DJI", "SP500": "^GSPC", "NASDAQ": "^IXIC",
    "USDINR": "INR=X", "CRUDE": "CL=F", "GOLD": "GC=F",
}


def to_yahoo_symbol(ticker: str) -> str:
    if not ticker:
        return ""
    raw = str(ticker).upper().strip()
    if raw.startswith("^"):
        return raw
    key = re.sub(r"[\s\-_&]", "", raw)
    if key in SPECIAL_MAP:
        return SPECIAL_MAP[key]
    if "=" in raw or raw.endswith(".NS") or raw.endswith(".BO"):
        return raw
    cleaned = raw.replace("&", "_")
    return f"{cleaned}.NS"


class YahooError(Exception):
    pass


@retry(
    stop=stop_after_attempt(3),
    wait=wait_exponential(multiplier=1, min=1, max=5),
    retry=retry_if_exception_type((httpx.TransportError, YahooError)),
    reraise=True,
)
async def _yahoo_get(path: str) -> Dict[str, Any]:
    settings = get_settings()
    url = f"{settings.yahoo_host}{path}"
    async with httpx.AsyncClient(timeout=15.0) as client:
        r = await client.get(
            url,
            headers={"User-Agent": "Mozilla/5.0 (compatible; DeskBackend/0.1)"},
        )
        if r.status_code >= 500:
            raise YahooError(f"yahoo {r.status_code}")
        if r.status_code >= 400:
            raise YahooError(f"yahoo {r.status_code}: {r.text[:200]}")
        return r.json()


async def get_quote(ticker: str) -> Dict[str, Any]:
    symbol = to_yahoo_symbol(ticker)
    data = await _yahoo_get(f"/v8/finance/chart/{symbol}?range=5d&interval=1d")
    result = (data.get("chart") or {}).get("result", [None])[0]
    if not result:
        raise YahooError(f"no data for {ticker}")
    meta = result.get("meta") or {}
    price = meta.get("regularMarketPrice")
    prev = meta.get("chartPreviousClose") or meta.get("previousClose") or price
    change = (price - prev) if price is not None and prev is not None else None
    change_pct = (change / prev * 100) if change is not None and prev else None
    return {
        "ticker": ticker.upper(),
        "symbol": symbol,
        "name": meta.get("longName") or meta.get("shortName") or ticker.upper(),
        "price": price,
        "previousClose": prev,
        "change": change,
        "changePct": change_pct,
        "currency": meta.get("currency") or "INR",
        "dayHigh": meta.get("regularMarketDayHigh"),
        "dayLow": meta.get("regularMarketDayLow"),
        "volume": meta.get("regularMarketVolume"),
        "exchange": meta.get("exchangeName"),
        "regularMarketTime": meta.get("regularMarketTime"),
    }


async def get_history(ticker: str, range_: str = "5y", interval: str = "1mo") -> List[Dict[str, Any]]:
    symbol = to_yahoo_symbol(ticker)
    data = await _yahoo_get(f"/v8/finance/chart/{symbol}?range={range_}&interval={interval}")
    result = (data.get("chart") or {}).get("result", [None])[0]
    if not result:
        raise YahooError(f"no history for {ticker}")
    timestamps = result.get("timestamp") or []
    closes = ((result.get("indicators") or {}).get("quote") or [{}])[0].get("close") or []
    out = []
    for t, c in zip(timestamps, closes):
        if c is not None:
            out.append({"date": _iso_from_epoch(t), "close": float(c)})
    return out


def _iso_from_epoch(ts: int) -> str:
    from datetime import datetime, timezone
    return datetime.fromtimestamp(ts, tz=timezone.utc).date().isoformat()
